### ISSUE TITLE
Bug: [no-unnecessary-type-parameters] False positive on curried functions

### ISSUE THREAD

**Issue by eranhirsch** [typescript-eslint/typescript-eslint#9721]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.4&showAST=types&fileType=.ts&code=C4TwDgpgBAYg9nAPAFQDRQMoD4oF5NQQAewEAdgCYDOUARFXALYQBGcFItAsAFBRQB%2BKMkIly1OnDIBjCNz6C6wOABsK8-gC46zDVG1kIANwgAnANy9eFCNJUBDU9ABmAVxnAAllKjOEKLAAKCntge21kAEptRGxAuGAACzNtDEi8HHgkNEwsSx4gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
type Foo<T, S> = S extends "somebody"
  ? T extends "once"
  ? "told"
  : "me"
  : never;

declare function foo<T>(data: T): <S>(other: S) => Foo<T, S>;
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-parameters": "error",
  },
};
```

### Expected Result

T shouldn't be surfaced by this rule; it has a meaningful usage in the type definitions.

### Actual Result

T is surfaced as only being used once, incorrectly.

### Additional Info

I maintain the [Remeda typescript utility library](https://github.com/remeda/remeda). We rely heavily on curried functions, which are functions that return functions. The rule is firing for almost all our complex types where the type is passed on to the return type of the curried function. The non-curried functions, which are identical in terms of how the type param is used, are not falsely detected, so this has something to do with the param being passed on "through" to the return type of its returned function type.

When trying to reproduce this, I noticed that simpler cases where the T is used directly in the type aren't falsely detected; the issue can only be reproduced once we start introducing ternary conditions on other type params.

You can see a fuller example via the PR, which tries to introduce this rule into our codebase: https://github.com/remeda/remeda/pull/801

**Comment by JoshuaKGoldberg** [comment]
👍 good call on the conditional types, thanks for the note. Marking as accepting PRs with an added appreciation for the _excellent_ Smash Mouth reference. 👌

**Comment by JoshuaKGoldberg** [comment]
...this is a funky one. 
* AST check: `context.sourceCode.getScope`'s `references` only lists the `T` in `data: T`, not `Foo<T, S>`.
* Types check:
  * For type `Foo<T, S>`, `type.checkType` is `S` and `type.extendsType` is `"somebody"`

**Comment by JoshuaKGoldberg** [comment]
...this is a funky one. 
* AST check: `context.sourceCode.getScope`'s `references` only lists the `T` in `data: T`, not `Foo<T, S>`.
* Types check:
  * For type `Foo<T, S>`, `type.checkType` is `S` and `type.extendsType` is `"somebody"`

Conditional types are _fun_ to work with.

**Comment by bradzacher** [comment]
> AST check: context.sourceCode.getScope's references only lists the T in data: T, not Foo<T, S>.

[In the playground](https://typescript-eslint.io/play/#ts=5.5.4&showAST=scope&fileType=.ts&code=C4TwDgpgBAYg9nAPAFQDRQMoD4oF5NQQAewEAdgCYDOUARFXALYQBGcFItAsAFBRQB%2BKMkIly1OnDIBjCNz6C6wOABsK8-gC46zDVG1kIANwgAnANy9eFCNJUBDU9ABmAVxnAAllKjOEKLAAKCntge21kAEptRGxAuGAACzNtDEi8HHgkNEwsSx4gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false) I can see both references:

![screenshot of scope analysis](https://github.com/user-attachments/assets/62913a30-b768-4e13-823f-4dd3be46982f)
